File: python/rational-numbers/rational_numbers.py
from __future__ import division

def reduce(numer, denom):
    n = 2
    while n <= min(abs(numer), abs(denom)):
        while(abs(numer) % n == 0 and abs(denom) % n == 0):
            numer //= n
            denom //= n
        n += 1
    if numer == 0:
        denom = 1
    if denom < 0:
        denom = abs(denom)
        numer = -numer
    return (numer, denom)

class Rational(object):
    def __init__(self, numer, denom):
        reduced_rational = reduce(numer, denom)
        self.numer = reduced_rational[0]
        self.denom = reduced_rational[1]

    def __eq__(self, other):
        return self.numer == other.numer and self.denom == other.denom

    def __repr__(self):
        return f'{self.numer}/{self.denom}'

    def __add__(self, other):
        numer = (self.numer * other.denom) + (other.numer * self.denom)
        denom = self.denom * other.denom
        return Rational(numer, denom)
        
    def __sub__(self, other):
        numer = (self.numer * other.denom) - (other.numer * self.denom)
        denom = self.denom * other.denom
        return Rational(numer, denom)

    def __mul__(self, other):
        numer = self.numer * other.numer
        denom = self.denom * other.denom
        return Rational(numer, denom)

    def __truediv__(self, other):
        return self * Rational(other.denom, other.numer)

    def __abs__(self):
        return Rational(abs(self.numer), abs(self.denom))

    def __pow__(self, power):
        numer = self.numer ** power
        denom = self.denom ** power
        return Rational(numer, denom)

    def __rpow__(self, base):
        return base ** (self.numer / self.denom)

File: typescript/rational-numbers/rational-numbers.ts
const GCD = (a: number, b: number): number => {
  a = Math.abs(a)
  b = Math.abs(b)
  const least = Math.min(a, b)
  const greatest = Math.max(a, b)
  let gcd = 1;
  for (let i = least; i > 1; i--) {
    if (least % i === 0 && greatest % i === 0) {
      gcd = i
      break;
    }
  }
  return gcd
}

export default class Rational {
  a: number
  b: number

  constructor(numerator: number, denominator: number) {
    this.a = numerator
    this.b = denominator
  }

  reduce(): Rational {
    if (this.a === 0) {
      return new Rational(0, 1)
    }
    if (this.b < 0) {
      this.a *= -1
      this.b *= -1
    }
    const gcd = GCD(this.a, this.b)
    return new Rational(this.a / gcd, this.b / gcd)
  }

  add(other: Rational): Rational {
    const a = this.a * other.b + other.a * this.b
    const b = this.b * other.b
    return new Rational(a, b).reduce()
  }

  sub(other: Rational) {
    return new Rational(this.a * other.b - other.a * this.b, this.b * other.b).reduce()
  }

  mul(other: Rational) {
    return new Rational(this.a * other.a, this.b * other.b).reduce()
  }

  div(other: Rational) {
    return new Rational(this.a * other.b, other.a * this.b).reduce()
  }

  abs(): Rational {
    return new Rational(Math.abs(this.a), Math.abs(this.b)).reduce()
  }

  exprational(pow: number): Rational {
    return new Rational(Math.pow(this.a, pow), Math.pow(this.b, pow)).reduce()
  }

  expreal(n: number): number {
    return Number(Math.pow(n, (this.a / this.b)).toPrecision(15))
  }
}